### ISSUE TITLE
CPU backend: profile and optimize PIConGPU

### ISSUE THREAD

**Issue by psychocoderHPC** [ComputationalRadiationPhysics/picongpu#2280]
It would be usefull to profile PIConGPU to find the current bottle neck methods.

# possible tool for CPU

- `gprof` : 
  - add compiler flag `-pg `
  - `./picongpu -d 1 1 1 -g 64 64 64  -s 50 --periodic 1 1 1 -p 1`
  - `gprof ./picongpu gmon.out > prof.txt`
- clone [grpof2dot] (https://github.com/jrfonseca/gprof2dot)
  - `cat ./prof.txt | gprof2dot.py -s  | dot -Tpng -o out.png`

This result is created on my laptop with `Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz` and the PR #2275 

![out](https://user-images.githubusercontent.com/4037125/30847908-864b81c4-a29d-11e7-80a1-d8837140552c.png)

**Comment by BenjaminW3** [comment]
I think what can be seen in the attached image is that allocating memory with new is slow, but we all knew this before ;-)
For each variable in block shared memory, a single allocation is performed. 
As far as I can imagine, CUDA compilers have an advantage here, because they could sum up the size of all block shared memory at compile time.
I will create a optimization ticket in alpaka.

**Comment by psychocoderHPC** [comment]
@BenjaminW3 You are right, but I am not sure if this is really the bottleneck. I played last week  around and changed the allocation to a single allocation of e.g. 64kb and than increment only the byte pointer each time a shared variable is allocated. This has not increased the performance. But this was only a 5 minute hack and I need to validate it a little bit better.

**Comment by psychocoderHPC** [comment]
After adding [this](https://github.com/ComputationalRadiationPhysics/alpaka/issues/409#issuecomment-332465165) patch I get a very performance boost of 4% on my laptop.

Profile picture with patch:

![out](https://user-images.githubusercontent.com/4037125/30910150-e0d9f788-a383-11e7-82da-cdf0664148c8.png)
